File: python_ruby workspace/inheritance_cal_2.py
class Cal(object):
    _history =[]
    def __init__(self, v1, v2):
        if isinstance(v1,int):
            self.v1=v1 
        if isinstance(v2,int):
            self.v2=v2
    def add(self):
        result = self.v1+self.v2
        Cal._history.append("add : %d+%d=%d" % (self.v1, self.v2, result))
        return result
    def subtract(self):
        result =  self.v1-self.v2
        Cal._history.append("subtract : %d-%d=%d" % (self.v1, self.v2, result))
        return result
    def setV1(self,v1):
        if isinstance(v1,int):
            self.v1=v1
    def getV1(self):
        return self.v1
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)

class Calmultiply(): 
    #기본 계산기능 상속에다가 곱하기 추가
    #생성자는 부모 클래스 따라감
    def multiply(self):
        result = self.v1*self.v2
        Cal._history.append("multiply : %d*%d=%d" % (self.v1, self.v2, result))
        return result


c1 = Calmultiply(10,20)
print(c1.add())
print(c1.multiply())

class Caldivide(Cal, Calmultiply):
        #순차 상속이 아닌 다중상속 처리함(후속 연습)
    def divide(self):
        result = self.v1/self.v2        
        Cal._history.append("divide : %d/%d=%d" % (self.v1, self.v2, result))
        return result
    
c2=Caldivide(20,4)
print(c2.add())
print(c2.multiply())
print(c2.divide())
Cal.history()
#위에꺼 다 상속해서 값 잘 나옴

File: python_ruby workspace/inheritance_cal_2.rb
class Cal
    @@_history = []
    def initialize(v1,v2) #여기서 V1은 지역변수임
        @v1=v1 #골뱅이는 인스턴스 변수임
        @v2=v2
    end
    def add()
        result = @v1+@v2
        @@_history.push("add : #{@v1}+#{@v2}=#{result}")
        return result
    end
    def subtract()
        result = @v1-@v2
        @@_history.push("subtract : #{@v1}-#{@v2}=#{result}")
        return result
    end
    def setV1(v1)
        if v1.is_a?(Integer)
        @v1=v1
        end
    end
    def getV1()
        return @v1
    end
    def Cal.history()
        for item in @@_history 
            p item
        end
    end
end

class Calmultiply < Cal
    #기본 계산기능 상속에다가 곱하기 추가
    #생성자는 부모 클래스 따라감    
    def multiply()
        result = @v1*@v2
        @@_history.push("multiply : #{@v1}*#{@v2}=#{result}")
        return result
    end
end

c1=Calmultiply.new(10,20)
p c1.add()
p c1.multiply()
p c1.subtract()

class Caldivide < Calmultiply
    #곱하기 추가된 기능 상속에다가 나누기 추가
    #생성자는 부모 클래스(곱하기) > 또 부모(계산기) 따라감
    def divide()
        result = @v1/@v2
        @@_history.push("divide : #{@v1}/#{@v2}=#{result}")
        return result
    end
end

c2=Caldivide.new(20,4)
p c2.add()
p c2.multiply()
p c2.divide()
Cal.history()
#위에꺼 다 상속해서 값 잘 나옴